src/js/dynload.ts

```diff
@@ -110,4 +140,84 @@ export async function loadDynlib(lib: string, global: boolean) {
   }
 }
 
+/**
+ * Load dynamic libraries inside a package.
+ *
+ * This function handles some painful details of loading dynamic libraries:
+ * - We need to load libraries in the correct order considering dependencies.
+ * - We need to load libraries globally if they are required by other libraries.
+ * - We need to tell Emscripten where to search for libraries.
+ * - The dynlib metadata inside a wasm module only contains the library name, not the path.
+ *   So we need to handle them carefully to avoid loading the same library twice.
+ *
+ * @param pkg The package metadata
+ * @param dynlibPaths The list of dynamic libraries inside a package
+ * @private
+ */
+export async function loadDynlibsFromPackage(
+  pkg: PackageData,
+  dynlibPaths: string[],
+) {
+  // assume that shared libraries of a package are located in <package-name>.libs directory,
+  // following the convention of auditwheel.
+  const auditWheelLibDir = `${API.sitepackages}/${
+    pkg.file_name.split("-")[0]
+  }.libs`;
+
+  // This prevents from reading large libraries multiple times.
+  const readFileMemoized: ReadFileType = memoize(Module.FS.readFile);
+
+  const globalLibs: Set<string> = calculateGlobalLibs(
+    dynlibPaths,
+    readFileMemoized,
+  );
+
+  const dynlibs = dynlibPaths.map((path) => {
+    const global = globalLibs.has(Module.PATH.basename(path));
+    return {
+      path: path,
+      global: global || !!pkg.shared_library,
+    };
+  });
+
+  // Sort libraries so that global libraries can be loaded first.
+  // TODO(ryanking13): It is not clear why global libraries should be loaded first.
+  //    But without this, we get a fatal error when loading the libraries.
+  dynlibs.sort((lib: { global: boolean }) => (lib.global ? -1 : 1));
+
+  for (const { path, global } of dynlibs) {
+    await loadDynlib(path, global, [auditWheelLibDir], readFileMemoized);
+  }
+}
+
+/**
+ * Given a list of libraries, return a list of libraries to load globally.
+ * @param libs The list of path to libraries
+ * @param readFileFunc A function to read the file, if not provided, use Module.FS.readFile
+ * @returns A list of libraries needed to be loaded globally
+ * @private
+ */
+function calculateGlobalLibs(
+  libs: string[],
+  readFileFunc?: ReadFileType,
+): Set<string> {
+  let readFile: ReadFileType = Module.FS.readFile;
+  if (readFileFunc !== undefined) {
+    readFile = readFileFunc;
+  }
+
+  const globalLibs = new Set<string>();
+
+  libs.forEach((lib: string) => {
+    const binary = readFile(lib);
+    const needed = Module.getDylinkMetadata(binary).neededDynlibs;
+    needed.forEach((lib: string) => {
+      globalLibs.add(lib);
+    });
```

Comment:
Is this a conservative overestimate of the libs that need to be loaded globally? Or is this exactly the condition? Also, does this code also run for shared libraries that are loaded separately? It's fairly efficient to turn an originally local library into a global one when it turns out to be needed later. Do we handle this case?

In any case, it's really nice to see these improvements.

Reply:
> Is this a conservative overestimate of the libs that need to be loaded globally? Or is this exactly the condition?

Since `neededDynlibs` contains all dynamic libraries linked during the compilation (via `-l` flag), I think this is not an overestimate unless we find a way to load symbols of external libraries loaded locally.

On the other hand, this can rather be an "underestimate" of libs, because users who write compile commands might omit to specify dynamic libraries that should be linked. For example: in #3234, I found and fixed that our `sharedlib-test-py` package [wasn't linked to `sharedlib-test.so` correctly](https://github.com/pyodide/pyodide/pull/3234/files#diff-ac3008498f6901cf8542ab2b87911b5477946de93a7babe0ded3279c912887f7).

> Also, does this code also run for shared libraries that are loaded separately?

Good point, I think we can skip this part for shared libraries that are loaded separately.

> It's fairly efficient to turn an originally local library into a global one when it turns out to be needed later. Do we handle this case?

No, I think we never turn a local library into a global one for now. Since `global` flag is propagated to dependencies when calling `loadDynamicLibrary`, a locally loaded dynamic library does not load its dependencies globally. 

(I'm not sure this behavior is a bug in Emscripten, probably I should open an issue in Emscripten).

Reply:
> (I'm not sure this behavior is a bug in Emscripten, probably I should open an issue in Emscripten).

Opened an issue in Emscripten (https://github.com/emscripten-core/emscripten/issues/18264). If it is accepted, it might be possible for us to remove whole global checks.